File: packages/i18nify-js/src/modules/core/__tests__/unflattenKeys.test.ts
import unflattenKeys from '../unflattenKeys';

jest.mock('../../../common/errorBoundary', () => ({
  withErrorBoundary: jest.fn((fn) => fn),
}));

describe('unflattenKeys', () => {
  it('rebuilds a nested object with the default delimiter', () => {
    expect(
      unflattenKeys({
        'user.name': 'Ada',
        'user.address.city': 'London',
      }),
    ).toEqual({
      user: {
        name: 'Ada',
        address: {
          city: 'London',
        },
      },
    });
  });

  it('supports a custom delimiter', () => {
    expect(
      unflattenKeys(
        {
          'user/profile/city': 'Mumbai',
        },
        { delimiter: '/' },
      ),
    ).toEqual({
      user: {
        profile: {
          city: 'Mumbai',
        },
      },
    });
  });

  it('replaces non-object intermediates while rebuilding nested values', () => {
    expect(
      unflattenKeys({
        user: 'Ada',
        'user.name': 'Grace',
      }),
    ).toEqual({
      user: {
        name: 'Grace',
      },
    });
  });

  it('throws when the input is not a plain object', () => {
    expect(() =>
      unflattenKeys([] as unknown as Record<string, unknown>),
    ).toThrow('unflattenKeys: input must be a plain object.');
  });

  it('throws when the delimiter is empty', () => {
    expect(() =>
      unflattenKeys({ 'user.name': 'Ada' }, { delimiter: '' }),
    ).toThrow('unflattenKeys: delimiter must be a non-empty string.');
  });
});

File: packages/i18nify-go/modules/core/unflatten_keys_test.go
package core

import (
	"reflect"
	"testing"
)

func TestUnflattenKeys(t *testing.T) {
	tests := []struct {
		name    string
		input   map[string]any
		opts    *UnflattenOptions
		want    map[string]any
		wantErr bool
	}{
		{
			name: "rebuilds nested maps with default delimiter",
			input: map[string]any{
				"user.name":         "Ada",
				"user.address.city": "London",
			},
			want: map[string]any{
				"user": map[string]any{
					"name": "Ada",
					"address": map[string]any{
						"city": "London",
					},
				},
			},
		},
		{
			name: "supports custom delimiter",
			input: map[string]any{
				"user/profile/city": "Mumbai",
			},
			opts: &UnflattenOptions{Delimiter: "/"},
			want: map[string]any{
				"user": map[string]any{
					"profile": map[string]any{
						"city": "Mumbai",
					},
				},
			},
		},
		{
			name: "replaces non-map intermediates while rebuilding nested values",
			input: map[string]any{
				"user":      "Ada",
				"user.name": "Grace",
			},
			want: map[string]any{
				"user": map[string]any{
					"name": "Grace",
				},
			},
		},
		{
			name:    "returns error for nil input",
			wantErr: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := UnflattenKeys(tt.input, tt.opts)
			if tt.wantErr {
				if err == nil {
					t.Fatal("expected an error but got nil")
				}
				return
			}
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			if !reflect.DeepEqual(got, tt.want) {
				t.Fatalf("UnflattenKeys() = %#v, want %#v", got, tt.want)
			}
		})
	}
}
